app/tests/spec/lib/metrics.js

```diff
@@ -112,95 +114,233 @@ function (chai, $, p, Metrics, AuthErrors, WindowMock, TestHelpers) {
     });
 
     describe('flush', function () {
-      var sentData, serverError;
-
-      function ajaxMock(options) {
-        sentData = options.data;
-
-        if (serverError) {
-          return p.reject({ statusText: serverError });
-        }
-
-        return p({});
-      }
+      var sandbox, xhr, environment;
 
       beforeEach(function () {
         metrics.destroy();
 
+        sandbox = sinon.sandbox.create();
+        xhr = { ajax: function () {} };
+        environment = new Environment(windowMock);
         metrics = new Metrics({
-          ajax: ajaxMock,
+          xhr: xhr,
           window: windowMock,
-          inactivityFlushMs: 100
+          inactivityFlushMs: 100,
+          environment: environment
         });
+
         metrics.init();
+        metrics.logEvent('foo');
+        metrics.logEvent('bar');
+      });
 
-        sentData = serverError = null;
+      afterEach(function () {
+        sandbox.restore();
       });
 
-      it('sends filtered data to the server and clears the event stream', function () {
-        metrics.logEvent('event1');
-        metrics.logEvent('event2');
+      describe('has sendBeacon', function () {
+        beforeEach(function () {
+          sandbox.stub(environment, 'hasSendBeacon', function () {
+            return true;
+          });
+        });
+
+        describe('sendBeacon succeeds', function () {
+          var result;
+
+          beforeEach(function (done) {
+            sandbox.stub(windowMock.navigator, 'sendBeacon', function () {
+              return true;
+            });
+            metrics.flush().then(function (r) {
+              result = r;
+              done();
+            }, done);
+          });
+
+          afterEach(function () {
+            result = undefined;
+          });
+
+          it('calls sendBeacon correctly', function () {
+            assert.isTrue(windowMock.navigator.sendBeacon.calledOnce);
+            assert.lengthOf(windowMock.navigator.sendBeacon.getCall(0).args, 2);
+            assert.strictEqual(windowMock.navigator.sendBeacon.getCall(0).args[0], '/metrics');
+            var data = JSON.parse(windowMock.navigator.sendBeacon.getCall(0).args[1]);
```

Comment:
Can you add a space above this line to help give an indication that the next section is logically one unit?